docarray/array/abstract_array.py

```diff
@@ -1,37 +1,94 @@
 from abc import abstractmethod
-from typing import TYPE_CHECKING, Dict, Iterable, List, Optional, Sequence, Type, Union
+from typing import TYPE_CHECKING, Generic, List, Sequence, Type, TypeVar, Union
 
 from docarray.document import BaseDocument
+from docarray.typing.abstract_type import AbstractType
 
 if TYPE_CHECKING:
+    from docarray.proto import DocumentArrayProto, NodeProto
     from docarray.typing import NdArray, TorchTensor
 
 
-class AbstractDocumentArray(Sequence):
+T = TypeVar('T', bound='AbstractDocumentArray')
+T_doc = TypeVar('T_doc', bound=BaseDocument)
 
+
+class AbstractDocumentArray(Sequence[BaseDocument], Generic[T_doc], AbstractType):
```

Comment:
What are the implications of not inheriting from list?

Reply:
I decided not to inherit from list because we cannot really follow the list interface, example the __class_getitem__ is clashing with the list one etcc. Overall it was a pain to follow the list interface because we slightly need to change it

Reply:
Makes sense, agree